ICAFIX/scripts/run_icasso.m

```diff
@@ -0,0 +1,253 @@
+%   nClust  : Number of icasso repetitions (default = '1')
+%   maxIter : Maximum number of iterations per ica fit (default = '1000')
+%
+% example inputs:
+% Dim = '41';
+% concatfmri = 'tfMRI_EMOTION_RL_LR';
+% concatfmrihp = 'tfMRI_EMOTION_RL_LR_hp0';
+% ConcatFolder = '/mnt/myelin/burke/HCPpipelines/dev_study/100307/MNINonLinear/Results/tfMRI_EMOTION_RL_LR_ICASSO';
+% vis = 'basic';
+% nICA = '2@2';
+% nClust = '2';
+% maxIter = '1000';
+% 
+% Dim = str2double(Dim);
+% nICA = cellfun(@str2double,regexp(nICA,'@','split'));
+% nClust = str2double(nClust);
+% maxIter = str2double(maxIter);
+
+% Created 2024-06-05
+% Burke Rosen
+
+% ToDo:
+% () Currently the vnts file and brainMaskFile paths and names are inferred from
+%    ConcatFolder, concatfmrihp, and concatfmri. Maybe they should be their own 
+%    arguments for flexibility.
+
+%% parse parameters
+if nargin < 4 || any(cellfun(@isempty,{Dim,concatfmri,concatfmrihp,ConcatFolder}))
+  error('Dim, concatfmri, concatfmrihp, and ConcatFolder are required!')
+end
+Dim = str2double(Dim);
+if nargin < 5 || isempty(vis) 
+  vis = 'basic';
+end
+if ~ismember(vis,{'basic','off'})
+  warning('vis must be ''basic'' or ''off'', reverting to ''basic''')
+  vis = 'basic';
+end
+if nargin < 6 || isempty(nICA) 
+  nICA = 100;
+else
+  nICA = cellfun(@str2double,regexp(nICA,'@','split'));
+end
+if nargin < 7 || isempty(nClust) 
+  nClust = 1;
+else
+  nClust = str2double(nClust);
+end
+if numel(nICA) ~= nClust
+  error('nICA must be a @ delimited list with a number of elements equal to nClust')
+end
+if nargin < 8 || isempty(maxIter) 
+  maxIter = 1000;
+else
+  maxIter = str2double(maxIter);
+end
+
+%% check IO dependencies
+% use imaging processing toolbox utilities, or if those are not available use FSL utilities 
+function out = out2(fun);[~,out] = fun();end
+function out = out3(fun);[~,~,out] = fun();end
+if isempty(which('niftiread'))
+  infoNIFTI = @(fName) struct('ImageSize',out2(@() read_avw(fName))','PixelDimensions',out3(@() read_avw(fName))');
+  readNIFTI = @(fName) read_avw(fName);
+  writeNIFTI = @(img,fName,hdr) save_avw(img,fName,'f',hdr.PixelDimensions);
+  if isempty(which('read_avw'))
+    error('neither niftiread nor read_avw on matlab path!')
+  end
+else
+  readNIFTI = @(fName) niftiread(fName);
+  infoNIFTI = @(fName) niftiinfo(fName);
+  writeNIFTI = @(img,fName,hdr) niftiwrite(img,fName,hdr);
+end
+
+%% parse paths
+% inputs
+vntsFile = sprintf('%s/%s_vnts.nii.gz',ConcatFolder,concatfmrihp);
+brainMaskFile = sprintf('%s/%s_brain_mask.nii.gz',ConcatFolder,concatfmri);
+if ~exist(vntsFile,'file');error('%s doesn''t exist!',vntsFile);end
+if ~exist(brainMaskFile,'file');error('%s doesn''t exist!',brainMaskFile);end
+
+% outputs
+outDir = sprintf('%s/%s.ica/filtered_func_data.ica',ConcatFolder,concatfmrihp);
+[~,~] = unix(['mkdir -p ' outDir]);
+
+%% load data, reshape to 2d, and apply brain mask
+vnts = double(readNIFTI(vntsFile));
```

Comment:
It would be better if the comment explained why it is being converted to double, since it is likely it was written to disk as single.  The masking and reshaping are somewhat evident from the code (though perhaps the function name could be improved to something like `maskAndSpatiallyFlatten`), and therefore are not as deserving of an explanatory comment.

Reply:
fastica needs double. But I typically cast data as double as soon as it is in memory to reduce any downstream rounding errors, and then covert back to single before storing again, unless I am RAM limited. It this script the placement of the recasting doesn't matter so long as it is before fastica. I added a comment indicating it is for fastica.

Reply:
function renamed.